src/uct/tcp/sockcm/sockcm_ep.c

```diff
@@ -14,26 +14,196 @@
+    pthread_mutex_init(&self->ops_mutex, NULL);
+    ucs_queue_head_init(&self->ops);
+
+    param_sockaddr = (struct sockaddr *) sockaddr->addr;
+    if (UCS_OK != ucs_sockaddr_sizeof(param_sockaddr, &sockaddr_len)) {
+       ucs_error("sockcm ep: unknown remote sa_family=%d", sockaddr->addr->sa_family);
+       status = UCS_ERR_IO_ERROR;
+       goto err;
+    }
+
+    memcpy(&self->remote_addr, param_sockaddr, sockaddr_len);
+
+    self->slow_prog_id = UCS_CALLBACKQ_ID_NULL;
+
+    status = uct_sockcm_ep_set_sock_id(self);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    status = ucs_sys_fcntl_modfl(self->sock_id_ctx->sock_id, O_NONBLOCK, 0); 
+    if (status != UCS_OK) {
+        goto sock_err;
+    }
+
+    status = ucs_socket_connect(self->sock_id_ctx->sock_id, param_sockaddr);
+    if (UCS_STATUS_IS_ERR(status)) {
+        ucs_debug("%d: connect fail\n", self->sock_id_ctx->sock_id);
+        self->conn_state = UCT_SOCKCM_EP_CONN_STATE_CLOSED;
+        goto sock_err;
+    }
+
+    self->conn_state = (status == UCS_INPROGRESS) ? 
+        UCT_SOCKCM_EP_CONN_STATE_SOCK_CONNECTING : 
+        UCT_SOCKCM_EP_CONN_STATE_SOCK_CONNECTED;
+
+    status = ucs_async_set_event_handler(iface->super.worker->async->mode,
+                                         self->sock_id_ctx->sock_id,
+                                         UCS_EVENT_SET_EVWRITE,
+                                         uct_sockcm_ep_event_handler,
+                                         self, iface->super.worker->async);
+    if (status != UCS_OK) {
+        goto sock_err;
+    }
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    ucs_list_add_tail(&iface->used_sock_ids_list, &self->sock_id_ctx->list);
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+
+    ucs_debug("created an SOCKCM endpoint on iface %p, "
+              "remote addr: %s", iface,
+               ucs_sockaddr_str(param_sockaddr,
+                                ip_port_str, UCS_SOCKADDR_STRING_LEN));
+    self->status = UCS_INPROGRESS;
     return UCS_OK;
+
+sock_err:
+    close(self->sock_id_ctx->sock_id);
+    ucs_free(self->sock_id_ctx);
+err:
+    ucs_debug("error in sock connect\n");
+    pthread_mutex_destroy(&self->ops_mutex);
+
+    return status;
 }
 
 static UCS_CLASS_CLEANUP_FUNC(uct_sockcm_ep_t)
 {
-}
+    uct_sockcm_iface_t *iface = NULL;
+
+    iface = ucs_derived_of(self->super.super.iface, uct_sockcm_iface_t);
+
+    ucs_debug("sockcm_ep %p: destroying", self);
+
+    ucs_async_remove_handler(self->sock_id_ctx->sock_id, 0);
+
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
 
+    uct_worker_progress_unregister_safe(&iface->super.worker->super,
+                                        &self->slow_prog_id);
+
+    pthread_mutex_destroy(&self->ops_mutex);
+    if (!ucs_queue_is_empty(&self->ops)) {
+        ucs_warn("destroying endpoint %p with not completed operations", self);
+    }
+
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
```

Comment:
where is sock_id_ctx destroyed?

Reply:
If sock_id_ctx isn't initialized correctly then here - https://github.com/openucx/ucx/pull/4073/files#diff-aca878e51b7d5e6eaa75cd20a1375eadR141

Otherwise here - 
https://github.com/openucx/ucx/pull/4073/files#diff-7cf03c5fefa318dcaeaf6bf7312d1911R255

Reply:
but if sockid is created for the endpoint, why delaying it's destruction until iface dtor?

Reply:
I had noticed that my earlier implementation of sockcm had some test hangs if I moved destruction to ep cleanup. I've moved it back and sockaddr tests are passing so hopefully there isn't a real issue in moving it back to ep cleanup.